### PR TITLE
Reduce storage costs

### LINKED ISSUE #2124: Minimise development costs
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a request for a new feature in the Data Safe Haven or an upgrade to an existing feature.
- [x] The feature is still missing in the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :strawberry: Suggested change

After the 5.0.0 release, we should look at our SRE costs and try to work out how to make development environments cheaper without compromising the performance of production environments.

## :steam_locomotive: How could this be done?

- reduce default storage sizes
- use cheaper VMs or disks
- reduce resources used by ACIs
- reduce Application Gateway costs
- reduce Firewall costs
- ...

### PR DESCRIPTION
**Pull request by jemrobinson**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

n/a

### :arrow_heading_up: Summary

Reduce storage costs by changing storage SKUs and quotas

- Reduce /home and /shared to the minimum size (100GiB)
    Screenshot 2024-08-16 at 13 35 19
- Add a config option to set the quota for /home and /shared
- Move config file share to transaction-optimised storage

### :closed_umbrella: Related issues

Progress towards #2124

### :microscope: Tests

Test details below:

TL;DR: This change reduces daily costs for an idle SRE from ~£70 to ~£45 an ~35% decrease.

#### Wednesday August 14th (before changes made)

Daily(!) cost summary of all SHMs/SREs

Screenshot 2024-08-16 at 12 39 29

Resources in `sre-coral` costing more than £1.50 per day:

Screenshot 2024-08-16 at 12 40 27

#### Sunday August 18th

**Note: this is a day when none of the SREs are being used: variations in costs will be due to different VM sizes/database deployments**

Screenshot 2024-08-19 at 09 42 39

Screenshot 2024-08-19 at 09 44 32

- User data is at ~10% of a comparable SRE
- Config data is at ~15% of a comparable SRE

### CODE REVIEW

File: data_safe_haven/types/annotated_types.py [hunk 1]
```diff
@@ -1,18 +1,20 @@
...
 from typing import Annotated, TypeAlias, TypeVar
 
+from annotated_types import Ge
 from pydantic import Field
 from pydantic.functional_validators import AfterValidator
 
 from data_safe_haven import validators
 
+AzureLocation = Annotated[str, AfterValidator(validators.azure_location)]
+AzurePremiumFileShareSize = Annotated[int, Ge(100)]
```
Comment (JimMadge) [inline data_safe_haven/types/annotated_types.py:11]: These annotated types are really helpful.

Comment (jemrobinson) [reply data_safe_haven/types/annotated_types.py:11]: Yeah - so much easier than writing your own validation logic.

**Review by JimMadge — COMMENTED** [review]
LGTM.

I think we should add a note to the docs about the default quota though. I imagine in practice it is quite small.

I'm thinking of users who have gone with the defaults then found the computing performance to be insufficient.

File: data_safe_haven/types/annotated_types.py (hunk 1 above)
Comment (JimMadge) [reply data_safe_haven/types/annotated_types.py:11]: :100:
